### PR TITLE
[SYCL] Adds support for device UUID as a SYCL extension.

### PR DESCRIPTION
**Pull request by rbegam**
This includes support for only level_zero. The
llvm/sycl/docs/extensions/IntelGPU/IntelGPUDeviceInfo.md doc has
been modified with a brief description of the extension. A new aspect
is added to indicate if the support is available.

Signed-off-by: rbegam <rehana.begam@intel.com>

### CODE REVIEW

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md [hunk 1]
```diff
@@ -11,7 +11,40 @@ This proposal details what is required to provide this information as a SYCL ext
 
 The Feature Test Macro will be defined as:
 
-    #define SYCL_EXT_INTEL_DEVICE_INFO 1
+    #define SYCL_EXT_INTEL_DEVICE_INFO 2
```
Comment (smaslov-intel) [inline sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:14]: Why did you change macro value to 2?

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md [hunk 2]
```diff
@@ -11,7 +11,40 @@ This proposal details what is required to provide this information as a SYCL ext
...
+
+
+# Device UUID #
+
+A new device descriptor will be added which will provide the device Universal Unique ID (UUID).
+
+This new device descriptor is only available for devices in the Level Zero platform, and the matching aspect is only true for those devices. The DPC++ default behavior is to expose GPU devices through the Level Zero platform.
```
Comment (smaslov-intel) [inline sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:21]: If we wanted this to be Level-Zero only then we'd add this extension to https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/LevelZeroBackend/LevelZeroBackend.md. My understanding is that the intention is to have it supported at least with OpenCL too, so let's not have this description be too Level-Zero centric.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1543,6 +1543,8 @@ pi_result piDeviceGetInfo(pi_device Device, pi_device_info ParamName,
     return ReturnValue(Device->Platform);
   case PI_DEVICE_INFO_VENDOR_ID:
     return ReturnValue(pi_uint32{Device->ZeDeviceProperties.vendorId});
+  case PI_DEVICE_INFO_UUID:
+    return ReturnValue(Device->ZeDeviceProperties.uuid.id);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1547]: This will return "uint8_t id[ZE_MAX_DEVICE_UUID_SIZE]" not the std::array. So let's "uin8_t *" be the PI interface.

File: sycl/include/CL/sycl/detail/pi.h [hunk 4]
```diff
@@ -64,6 +65,8 @@ using pi_uint64 = uint64_t;
 using pi_bool = pi_uint32;
 using pi_bitfield = pi_uint64;
 using pi_native_handle = uintptr_t;
+using pi_byte_array = std::array<std::byte, 16>;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:68]: I realized that this is not a good idea to use STL in PI API. The goal was to keep it C-only such that there are no C++ compatibility issues. Let's instead have the first call to know the size, have users allocate a C-array, and have the second call to populate it.

File: sycl/include/CL/sycl/detail/pi.h [hunk 5]
```diff
@@ -278,6 +281,8 @@ typedef enum {
       CL_DEVICE_CROSS_DEVICE_SHARED_MEM_CAPABILITIES_INTEL,
   PI_DEVICE_INFO_USM_SYSTEM_SHARED_SUPPORT =
       CL_DEVICE_SHARED_SYSTEM_MEM_CAPABILITIES_INTEL,
+  // Intel UUID extension.
+  PI_DEVICE_INFO_UUID = CL_DEVICE_UUID_KHR,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:285]: Is `CL_DEVICE_UUID_KHR` not supported by OpenCL RT, but is in the header?

File: sycl/include/CL/sycl/feature_test.hpp [hunk 6]
```diff
@@ -13,7 +13,7 @@ namespace sycl {
 // Feature test macro definitions
 
 // TODO: Move these feature-test macros to compiler driver.
-#define SYCL_EXT_INTEL_DEVICE_INFO 1
+#define SYCL_EXT_INTEL_DEVICE_INFO 2
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/feature_test.hpp:16]: Again, why incremented this to 2?

File: sycl/source/detail/device_impl.cpp [hunk 7]
```diff
@@ -284,6 +284,11 @@ bool device_impl::has(aspect Aspect) const {
                MDevice, PI_DEVICE_INFO_GPU_EU_COUNT_PER_SUBSLICE,
                sizeof(pi_device_type), &device_type,
                &return_size) == PI_SUCCESS;
+  case aspect::ext_intel_device_info_uuid:
+    return getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
```
Comment (smaslov-intel) [inline sycl/source/detail/device_impl.cpp:288]: here you would do 2 calls

**Review by smaslov-intel — APPROVED** [review]

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
Please address review comments

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 1 above)
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:14]: wanted to treat it as a versioning for the extensions added. any value >= 2 will have UUID support.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 2 above)
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:21]: ok, will modify.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1547]: ok, will modify.

File: sycl/include/CL/sycl/detail/pi.h (hunk 5 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/pi.h:285]: yes, that is the case at this point.

Comment (smaslov-intel) [reply sycl/include/CL/sycl/detail/pi.h:285]: OK

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 1 above)
Comment (smaslov-intel) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:14]: I don't think it was designed this way, but don't actually see why we couldn't use it this way. @gmlueck , what do you think?

Comment (s-kanaev) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:14]: As SYCL 2020 spec says (6.3.3 Feature test macros) this **is** the way it was designed:

> Since the interface to an extension might change from one release to another, vendors are also encouraged to predefine the macro’s value to the version of the extension. Vendors should use a numerical value that monotonically increases for each revision of the extension API.

Comment (smaslov-intel) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:14]: OK, great.

File: sycl/include/CL/sycl/feature_test.hpp (hunk 6 above)
Comment (rbegam) [reply sycl/include/CL/sycl/feature_test.hpp:16]: same as above

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md]: done as suggested. added a 'version' sub-section for each to specify the number though.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md [hunk 8]
```diff
@@ -9,9 +9,50 @@ This proposal details what is required to provide this information as a SYCL ext
...
+## Version ##
+
+The extension supports this query in version 2 and later.
+
+
+## Device Information Descriptors ##
+
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| info\:\:device\:\:ext\_intel\_device\_info\_uuid | uint8\_t \* | Returns the device UUID|
```
Comment (gmlueck) [inline sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:36]: Why did you change the return type here?  Returning `std::array<std::byte, 16>` was much better, I think.  With `uint8_t *`, it's unclear whether the caller is supposed to deallocate the memory.

File: sycl/include/CL/sycl/feature_test.hpp [hunk 9]
```diff
@@ -13,7 +13,7 @@ namespace sycl {
 // Feature test macro definitions
 
 // TODO: Move these feature-test macros to compiler driver.
```
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/include/CL/sycl/feature_test.hpp:15]: Yes, I still think we want the compiler driver to define these feature-test macros.  We believe our intention is to provide compiler driver options that enable/disable each extension individually.  If we do that, then the compiler driver will need to conditionally define these macros.

File: sycl/source/detail/device_impl.cpp [hunk 10]
```diff
@@ -284,6 +284,20 @@ bool device_impl::has(aspect Aspect) const {
...
+  case aspect::ext_intel_device_info_uuid: {
+    auto Result = getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
+        MDevice, PI_DEVICE_INFO_UUID, sizeof(pi_device_type), &device_type,
+        &return_size);
+    if (Result != PI_SUCCESS) {
+      return false;
+    }
+
+    pi_uint8_ptr uuid =
+        static_cast<pi_uint8_ptr>(malloc(return_size * sizeof(uint8_t)));
```
[the remark this answers is not in the document]
Comment (rbegam) [reply sycl/source/detail/device_impl.cpp:296]: thanks for pointing this out.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 8 above)
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:36]: @smaslov-intel pointed out that it's preferable to not use STLs to maintain C compatibility. Changing this does include the requirement for deallocation.

File: sycl/source/detail/device_info.hpp [hunk 11]
```diff
@@ -1045,45 +1045,53 @@ get_device_info_host<info::device::ext_intel_pci_address>() {
       PI_INVALID_DEVICE);
 }
 template <>
-inline cl_uint get_device_info_host<info::device::ext_intel_gpu_eu_count>() {
+inline pi_uint32 get_device_info_host<info::device::ext_intel_gpu_eu_count>() {
```
[the remark this answers is not in the document]
Comment (rbegam) [reply sycl/source/detail/device_info.hpp:1048]: will do.

File: sycl/source/detail/device_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (gmlueck) [reply sycl/source/detail/device_impl.cpp:296]: I assume that we will change the DPC++ interface back to `std::array` (see my other comment).  In that case, this code here will be a memory leak, which is bad.  If the DPC++ interface assumes that the UUID is always 16-bytes, then there is no point in trying to support an arbitrary size here.  Therefore, you could just allocate a fixed 16-byte array to receive the results from the PI query.  Something like:

```
assert(return_size <= 16);
std::byte buf[16];
/* read contents of "buf" from PI layer */
```

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 8 above)
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:36]: my bad, will fix it.

File: sycl/source/detail/device_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (smaslov-intel) [reply sycl/source/detail/device_impl.cpp:296]: Since PI interface readily supports receiving the size then  I'd still go with 2 calls. You can additionally assert here that the returned size is 16 since you'd need to truncate it to fit the array of 16-bytes returned in SYCL API.

File: sycl/include/CL/sycl/aspects.hpp [hunk 12]
```diff
@@ -38,7 +38,8 @@ enum class aspect {
   ext_intel_gpu_subslices_per_slice = 22,
   ext_intel_gpu_eu_count_per_subslice = 23,
   ext_intel_max_mem_bandwidth = 24,
-  ext_intel_mem_channel = 25
+  ext_intel_mem_channel = 25,
+  ext_intel_device_info_uuid = 26
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/aspects.hpp:42]:
```suggestion
  ext_intel_device_info_uuid = 26,
```
NIT. To avoid changing two lines next time.

File: sycl/include/CL/sycl/detail/pi.h [hunk 13]
```diff
@@ -52,6 +52,7 @@
 #include <CL/cl_ext.h>
 #include <CL/sycl/detail/cl.h>
 #include <CL/sycl/detail/export.hpp>
+#include <array>
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/pi.h:55]:
```suggestion
#include <CL/sycl/detail/export.hpp>

#include <array>
```
NIT

File: sycl/source/detail/device_impl.cpp [hunk 14]
```diff
@@ -284,6 +284,20 @@ bool device_impl::has(aspect Aspect) const {
                MDevice, PI_DEVICE_INFO_GPU_EU_COUNT_PER_SUBSLICE,
                sizeof(pi_device_type), &device_type,
                &return_size) == PI_SUCCESS;
+  case aspect::ext_intel_device_info_uuid: {
+    auto Result = getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
+        MDevice, PI_DEVICE_INFO_UUID, sizeof(pi_device_type), &device_type,
```
Comment (romanovvlad) [inline sycl/source/detail/device_impl.cpp:289]: Probably I'm missing something, but won't this call write UUID values to the `device_type` variable? If so, why?
Shouldn't 3rg argument be `0` and 4th `nullptr`?

File: sycl/source/detail/device_impl.cpp [hunk 15]
```diff
@@ -284,6 +284,20 @@ bool device_impl::has(aspect Aspect) const {
...
+        &return_size);
+    if (Result != PI_SUCCESS) {
+      return false;
+    }
+
+    pi_uint8_ptr uuid =
+        static_cast<pi_uint8_ptr>(malloc(return_size * sizeof(uint8_t)));
+    return getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
+               MDevice, PI_DEVICE_INFO_UUID, sizeof(pi_uint8_ptr), uuid,
+               &return_size) == PI_SUCCESS;
```
Comment (romanovvlad) [inline sycl/source/detail/device_impl.cpp:299]:
```suggestion
               nullptr) == PI_SUCCESS;
```
NIT

File: sycl/source/detail/device_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/source/detail/device_impl.cpp:296]: Sorry, could somebody please clarify the resolution here? Will return value of the SYCL device query be `std::array<uint8_t, 16>`?

File: sycl/include/CL/sycl/detail/pi.h (hunk 13 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/pi.h]: done

File: sycl/include/CL/sycl/aspects.hpp (hunk 12 above)
Comment (rbegam) [reply sycl/include/CL/sycl/aspects.hpp]: done

File: sycl/source/detail/device_impl.cpp (hunk 15 above)
Comment (rbegam) [reply sycl/source/detail/device_impl.cpp]: done.

File: sycl/source/detail/device_impl.cpp (hunk 14 above)
Comment (rbegam) [reply sycl/source/detail/device_impl.cpp:289]: here trying to get the size.

File: sycl/source/detail/device_impl.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (rbegam) [reply sycl/source/detail/device_impl.cpp]: it's std::array<std::byte, 16>

File: sycl/include/CL/sycl/info/info_desc.hpp [hunk 16]
```diff
@@ -22,6 +25,8 @@ class platform;
 // TODO: stop using OpenCL directly, use PI.
 namespace info {
 
+using array_type = std::array<std::byte, 16>;
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/info/info_desc.hpp:28]:
```suggestion
using uuid_type = std::array<std::byte, 16>;
```

The array_type is way to generic for this typedef

File: sycl/include/CL/sycl/detail/pi.h [hunk 17]
```diff
@@ -278,6 +279,8 @@ typedef enum {
       CL_DEVICE_CROSS_DEVICE_SHARED_MEM_CAPABILITIES_INTEL,
   PI_DEVICE_INFO_USM_SYSTEM_SHARED_SUPPORT =
       CL_DEVICE_SHARED_SYSTEM_MEM_CAPABILITIES_INTEL,
+  // Intel UUID extension.
+  PI_DEVICE_INFO_UUID = CL_DEVICE_UUID_KHR,
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/detail/pi.h:283]: In addition to this the value should be described for piDeviceGetInfo (and what it returns).
And also, please make sure that OpenCL/CUDA plugins would do smth meaningful for it.

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
LGTM overall. just few small comments.

File: sycl/source/detail/device_impl.cpp (hunk 14 above)
Comment (romanovvlad) [reply sycl/source/detail/device_impl.cpp:289]: It seems we should pass 0 and nullptr as 3rd and 4th arguments. Otherwise low level runtime will try to write N bytes to the memory pointed by &device_type.

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 16 above)
Comment (rbegam) [reply sycl/include/CL/sycl/info/info_desc.hpp:28]: We need this type as info/device_traits.def couldn't take ret_type param in this 'std::array<std::byte, 16>' format, needed a typedef as a workaround. I actually tried sycl::detail namespace first, but info/device_traits.def could only use def from sycl::info.

Comment (gmlueck) [reply sycl/include/CL/sycl/info/info_desc.hpp:28]: These seem like limitations of the tools that process ".def" files.  Can those limitations be removed?

File: sycl/source/detail/device_impl.cpp (hunk 14 above)
Comment (romanovvlad) [reply sycl/source/detail/device_impl.cpp:289]: @rbegam Could you please address this comment?

Comment (rbegam) [reply sycl/source/detail/device_impl.cpp:289]: did you mean `sizeof(pi_device_type)`, `&device_type` these two parameters? I mistakenly thought `&device_type` and `&return_size`. I think you are right. I am fixing this.

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 16 above)
Comment (rbegam) [reply sycl/include/CL/sycl/info/info_desc.hpp]: I've added this under sycl::detail, I might have missed something while I tried that before.

Comment (gmlueck) [reply sycl/include/CL/sycl/info/info_desc.hpp]: Nice!

**Comment by rbegam** [comment]
ping @smaslov-intel

**Review by smaslov-intel — APPROVED** [review]
LGTM. Please add a test too.

File: sycl/include/CL/sycl/detail/type_traits.hpp [hunk 18]
```diff
@@ -32,6 +35,11 @@ using half = detail::half_impl::half;
 template <typename ElementType, access::address_space Space> class multi_ptr;
 
 namespace detail {
+// Type for Intel device UUID extension.
+// For details about this extension, see
+// sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md
+using uuid_type = std::array<std::byte, 16>;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/type_traits.hpp:41]: It seems `std::byte` is C++17 feature. Could you please either replace it with `unsigned char` or guard with C++17 check?

File: sycl/source/detail/device_impl.cpp [hunk 19]
```diff
@@ -284,6 +284,20 @@ bool device_impl::has(aspect Aspect) const {
...
+        MDevice, PI_DEVICE_INFO_UUID, 0, nullptr, &return_size);
+    if (Result != PI_SUCCESS) {
+      return false;
+    }
+
+    assert(return_size <= 16);
+    std::byte uuid[16];
+
+    return getPlugin().call_nocheck<detail::PiApiKind::piDeviceGetInfo>(
+               MDevice, PI_DEVICE_INFO_UUID, 16 * sizeof(std::byte), uuid,
```
Comment (romanovvlad) [inline sycl/source/detail/device_impl.cpp:298]: `std::byte` needs to be either replaced or guarded here as well.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md [hunk 20]
```diff
@@ -9,9 +9,50 @@ This proposal details what is required to provide this information as a SYCL ext
...
+| ----- | ----------- |
+| 1     | Initial extension version\. Base features are supported |
+| 2     | Device UUID is supported |
+
+
+# Device UUID #
+
+A new device descriptor will be added which will provide the device Universal Unique ID (UUID).
+
+This new device descriptor is currently only available for devices in the Level Zero platform, and the matching aspect is only true for those devices. The DPC++ default behavior would be to expose the UUIDs of all supported GPU devices which enables detection of total number of unique devices.
```
Comment (romanovvlad) [inline sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:24]: Could you please clarify if there any reason to say "GPU" here? Won't this feature work for other device types?

File: sycl/source/detail/device_impl.cpp (hunk 19 above)
Comment (gmlueck) [reply sycl/source/detail/device_impl.cpp:298]: This file is part of the runtime implementation, not a header file that might be compiled into the user's application.  I'm just curious why we cannot use C++17 features inside our runtime implementation.

File: sycl/include/CL/sycl/detail/type_traits.hpp (hunk 18 above)
Comment (gmlueck) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: This is an interesting point.  I think the declaration here is part of the public API for the extension.  The Khronos SYCL group has adopted C++17 as the base language for SYCL, so we use C++17 features when it makes sense to do so.  This does seem like a place where it makes sense.

Do DPC++ features follow the same principle where we use C++17 features where it makes sense?  If so, then I guess we would guard this code with an #ifdef, which means the feature will not be available for customers who try to compile in C++14 mode?

Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: Yes, the idea is to guard it so the feature is available if an app is compiled with C++17 or later.
Tagging @pvchupin

File: sycl/source/detail/device_impl.cpp (hunk 19 above)
Comment (romanovvlad) [reply sycl/source/detail/device_impl.cpp:298]: Sorry, you are right, no need to guard in the library.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md (hunk 20 above)
Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:24]: AFAIK, there is no restriction that it won't work for other devices. I'll remove it. thanks.

Comment (gmlueck) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:24]: I thought this query is going to be implemented only for the Level Zero backend, which works only for GPU devices.  Are we planning to implement it on other backends too?

Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:24]: this will be on opencl as well.

File: sycl/include/CL/sycl/detail/type_traits.hpp (hunk 18 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: if the UUID feature is available **only** when compiled with C++17 or later, don't we need the guard on others?

Comment (gmlueck) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: > don't we need the guard on others?

Not sure what you mean.  Guard on other features?  This feature only needs the guard because it uses `std::byte` which is not available on compilers older than C++17.  Other features don't use `std::byte`.  I think there are a few other SYC features that use C++17 things, and those are also guarded.  (I'm not sure about this, though.)

Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: sorry for the confusion. I meant putting guard on the other code segs of this PR.

Comment (gmlueck) [reply sycl/include/CL/sycl/detail/type_traits.hpp:41]: I think the guard is only needed on code that exists in the public header files because those are compiled by the end users, and the end user might have an older C++ compiler.  We always compile DPC++ itself with C++17, so there is no need to guard code in the internals of the runtime because we know that compiler will always support C++17.

File: sycl/test/on-device/extensions/intel-ext-device.cpp [hunk 21]
```diff
@@ -100,6 +102,20 @@ int main(int argc, char **argv) {
               std::cout << "Failed!" << std::endl;
               return 1;
             }
+          }
+
+          if (SYCL_EXT_INTEL_DEVICE_INFO >= 2) {
+            if (dev.has(aspect::ext_intel_device_info_uuid)) {
+              auto uuid = dev.get_info<info::device::ext_intel_device_info_uuid>();
```
[the remark this answers is not in the document]
Comment (rbegam) [reply sycl/test/on-device/extensions/intel-ext-device.cpp]: working on a unittest.

File: sycl/include/CL/sycl/detail/type_traits.hpp (hunk 18 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp:42]: the guard is causing the buildbot fails for c++14: http://ci.llvm.intel.com:8010/#/builders/37/builds/9166

Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp:42]: what is the expected behavior here? @romanovvlad

Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/type_traits.hpp:42]: It seems uses of `uuid_type` need also be guarded or we need to have an else branch with an alternative implementation of `uuid_type`.

Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp:42]: another option could be using `unsigned char` instead of `std::byte` as `else` branch would have it any way?

**Review by smaslov-intel — APPROVED** [review]

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+} // namespace
+
+static std::unique_ptr<TestCtx> TestContext;
+
+static pi_result redefinedDeviceGetInfo(pi_device device,
+                                        pi_device_info param_name,
+                                        size_t param_value_size,
+                                        void *param_value,
+                                        size_t *param_value_size_ret) {
+  return PI_SUCCESS;
```
Comment (romanovvlad) [inline sycl/unittests/kernel-and-program/DeviceInfo.cpp:33]: Could you please return some result for `param_name` == `PI_DEVICE_INFO_UUID` and check this result on line 76?

Comment (rbegam) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:38]: done.

**Comment by rbegam** [comment]
ping @smaslov-intel @romanovvlad

**Review by smaslov-intel — DISMISSED** [review]

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp (hunk 22 above)
Comment (romanovvlad) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:38]: Sorry if I was not clear. I meant that the function could return some UUID, e.g. 0xFF, and this value can be checked in the main function. I'm OK if this improvement is done in a separate PR.

**Review by romanovvlad — APPROVED** [review]

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp (hunk 22 above)
Comment (rbegam) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:38]: the change did not got uploaded, my bad. I am uploading it now.

Comment (rbegam) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:38]: @romanovvlad please review now.

**Comment by rbegam** [comment]
could you please check my latest comment and review the unittest? Let me know if it's not visible. thanks. @romanovvlad

**Comment by rbegam** [comment]
@gmlueck could you please review? thanks.

File: sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md [hunk 23]
```diff
@@ -9,9 +9,50 @@ This proposal details what is required to provide this information as a SYCL ext
...
+## Version ##
+
+The extension supports this query in version 2 and later.
+
+
+## Device Information Descriptors ##
+
+| Device Descriptors | Return Type | Description |
+| ------------------ | ----------- | ----------- |
+| info\:\:device\:\:ext\_intel\_device\_info\_uuid | unsigned char | Returns the device UUID|
```
Comment (gmlueck) [inline sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:36]: It looks like this line changed since I last reviewed.  The return type is listed here as `unsigned char`, but that seems wrong.  Shouldn't this be `std::array<unsigned char, 16>` instead?

Comment (rbegam) [reply sycl/doc/extensions/IntelGPU/IntelGPUDeviceInfo.md:36]: good catch, thanks.. fixing it.

**Review by gmlueck — DISMISSED** [review]
LGTM

**Review by smaslov-intel — APPROVED** [review]
LGTM

File: sycl/include/CL/sycl/detail/type_traits.hpp [hunk 24]
```diff
@@ -15,6 +15,8 @@
 
 #include <type_traits>
 
+#include <array>
```
Comment (bader) [inline sycl/include/CL/sycl/detail/type_traits.hpp:18]: Style suggestion.
```suggestion
#include <array>
#include <type_traits>
```

File: sycl/include/CL/sycl/info/info_desc.hpp [hunk 25]
```diff
@@ -131,7 +131,8 @@ enum class device : cl_device_info {
   usm_shared_allocations = PI_USM_SINGLE_SHARED_SUPPORT,
   usm_restricted_shared_allocations = PI_USM_CROSS_SHARED_SUPPORT,
   usm_system_allocator = PI_USM_SYSTEM_SHARED_SUPPORT,
-
+  // uuid extension
+  ext_intel_device_info_uuid = PI_DEVICE_INFO_UUID,
   // intel extensions
```
Comment (bader) [inline sycl/include/CL/sycl/info/info_desc.hpp:136]: I think `ext_intel_device_info_uuid` value should go after `// intel extensions` comment. Probably it should be added at the end of intel extensions list.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 26]
```diff
@@ -1514,6 +1514,10 @@ pi_result cuda_piDeviceGetInfo(pi_device device, pi_device_info param_name,
   case PI_DEVICE_INFO_GPU_SUBSLICES_PER_SLICE:
   case PI_DEVICE_INFO_GPU_EU_COUNT_PER_SUBSLICE:
   case PI_DEVICE_INFO_MAX_MEM_BANDWIDTH:
+    // TODO: Check if Intel device UUID extension is utilized for CUDA.
```
Comment (bader) [inline sycl/plugins/cuda/pi_cuda.cpp:1517]: If so, shouldn't we make it `oneapi` extension instead of `intel`?

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp [hunk 27]
```diff
@@ -0,0 +1,89 @@
+//==-------------- DeviceInfo.cpp --- device info unit test ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#define SYCL2020_DISABLE_DEPRECATION_WARNINGS
```
Comment (bader) [inline sycl/unittests/kernel-and-program/DeviceInfo.cpp:9]: Do we need this? Can we avoid using deprecated APIs in new tests?

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+#include <CL/sycl.hpp>
+#include <detail/context_impl.hpp>
+#include <gtest/gtest.h>
+#include <helpers/PiMock.hpp>
+
+using namespace sycl;
+
+namespace {
+struct TestCtx {
+  TestCtx(context &Ctx) : Ctx{Ctx} {};
```
Comment (bader) [inline sycl/unittests/kernel-and-program/DeviceInfo.cpp:20]:
```suggestion
  TestCtx(context &Ctx) : Ctx{Ctx} {}
```

**Review by s-kanaev — DISMISSED** [review]
SYCL RT changes LGTM

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp (hunk 27 above)
Comment (rbegam) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:9]: will remove it.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 26 above)
Comment (rbegam) [reply sycl/plugins/cuda/pi_cuda.cpp:1517]: I am not sure if that was the intension of the jira, but I could be wrong. @smaslov-intel what do you think we should do?

Comment (smaslov-intel) [reply sycl/plugins/cuda/pi_cuda.cpp:1521]: I think we should keep it to be intel extension for now.

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 25 above)
Comment (bader) [reply sycl/include/CL/sycl/info/info_desc.hpp:136]: @rbegam, will you resolve this comment?

**Review by bader — DISMISSED** [review]
@rbegam, please, avoid forced pushed. It makes tracking conversations difficult.

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 25 above)
Comment (rbegam) [reply sycl/include/CL/sycl/info/info_desc.hpp:136]: missed it somehow. fixing it.

**Review by romanovvlad — APPROVED** [review]
LGTM

File: sycl/include/CL/sycl/info/info_desc.hpp (hunk 25 above)
Comment (rbegam) [reply sycl/include/CL/sycl/info/info_desc.hpp]: fixed it here.

**Review by romanovvlad — APPROVED** [review]

**Comment by bader** [comment]
I still see a bunch of unaddressed comments. What are we going to do with them?

File: sycl/include/CL/sycl/detail/type_traits.hpp (hunk 18 above)
Comment (rbegam) [reply sycl/include/CL/sycl/detail/type_traits.hpp]: we decided to go with `unsigned char`.

**Comment by rbegam** [comment]
@bader all those comments were discussed offline and resolved. I've marked them as so now.

File: sycl/unittests/kernel-and-program/DeviceInfo.cpp (hunk 22 above)
Comment (rbegam) [reply sycl/unittests/kernel-and-program/DeviceInfo.cpp:36]: upon discussion with vlad, we've decided to have the change on a diffPR.

**Review by smaslov-intel — APPROVED** [review]

**Comment by rbegam** [comment]
ping @gmlueck

**Review by gmlueck — APPROVED** [review]

**Comment by zejun-chen** [comment]
Hi, @rbegam 

I have a question. For the device UUID, is there any platform, whose backend is `sycl::backend::ext_oneapi_level_zero`, but doesn't support device UUID?

Thank you.
